lib/logstash/inputs/lumberjack.rb

```diff
@@ -63,41 +64,61 @@ def register
     @circuit_breaker = LogStash::CircuitBreaker.new("Lumberjack input",
                             :exceptions => [LogStash::SizedQueueTimeout::TimeoutError])
 
+    @mutex = Mutex.new
+    @acccepting_connection = ConditionVariable.new
   end # def register
 
   def run(output_queue)
     start_buffer_broker(output_queue)
+    start_accepting_connection
+    @mutex.synchronize { @acccepting_connection.wait(@mutex) while !stop? }
+  end # def run
+
+  def stop
+    super
 
-    while true do
-      # Wrappingu the accept call into a CircuitBreaker
-      if @circuit_breaker.closed?
-        connection = @lumberjack.accept # Blocking call that creates a new connection
-
-        invoke(connection, @codec.clone) do |_codec, line, fields|
-          _codec.decode(line) do |event|
-            begin
-              decorate(event)
-              fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
-              @circuit_breaker.execute { @buffered_queue.push(event, @congestion_threshold) }
-            rescue => e
-              raise e
+    @threadpool.kill
+    @lumberjack.close
```

Comment:
this is ruby-lumberjack's close right? can it raise an exception? looking at the code:

```ruby
def close
     @close.make_true
     @server.close unless @server.closed?
    end
```

we don't check for @server being nil, is that a possibility?